feat(docs) Adding some additional clarity around low power states (#2330)
zmkfirmware/zmk: docs/docs/features/low-power-states/low-power-states.mdx

Reviewer: If we rename to `index.mdx` can we use `features/low-power-states` in the sidebar?

Author: Unfortunately it doesn't seem like it, just did a simple test.

## Soft Off

The soft off feature is used to turn the keyboard on and off explicitly, rather than through a timeout like the deep sleep feature. Depending on the keyboard, this may be through a dedicated on/off push button defined in hardware, or merely through an additional binding in the keymap to turn the device off and an existing reset button to turn the device back on.

The feature is intended as an alternative to using a hardware switch to physically cut power from the battery to the keyboard. This can be useful for existing PCBs not designed for wireless that don't have a power switch, or for new designs that favor a push button on/off like found on other devices. It yields power savings comparable to the deep sleep state.

:::note

The power off is accomplished by putting the MCU into a "soft off" state. Power is _not_ technically removed from the entire system, but the device will only be woken from the state by a few possible events.

:::

A device can be put in the soft off state by:

- Triggering a hardware-defined dedicated GPIO pin, if one exists;

> - Triggering the [soft off Behavior](../../behaviors/soft-off.md) from the keymap.

Reviewer: ```suggestion
- Triggering the [soft off behavior](../../behaviors/soft-off.md) from the keymap.
```

Once in the soft off state, the device can only be woken up by:

- Triggering any GPIO pin specified to enable waking from sleep, if one exists;
- Pressing a reset button found on the device.

> The GPIO pin used to wake from sleep can be a hardware-defined one, such as for a dedicated on-off push button, or it can be a single specific keyswitch reused for waking up(may be pressed accidentally during transit). To allow the simultaneous pressing of multiple keyswitches to trigger and exit soft off, some keyboards make use of additional hardware to integrate the dedicated GPIO pin into the keyboard matrix.

Reviewer: ```suggestion
The GPIO pin used to wake from sleep can be a hardware-defined one, such as for a dedicated on-off push button, or it can be a single specific key switch reused for waking up (which may be accidentally pressed, e.g. while the device is being carried in a bag). To allow the simultaneous pressing of multiple keyswitches to trigger and exit soft off, some keyboards make use of additional hardware to integrate the dedicated GPIO pin into the keyboard matrix.
```

### Config

Soft off must be enabled via [its corresponding config](../../config/power.md#low-power-states) before it can be used.

### Using Soft Off

If your keyboard has hardware designed to take advantage of soft off, refer to your keyboard's documentation.

> For keyboards which do not have such hardware, using soft off is as simple as placing the [Soft Off Behavior](../../behaviors/soft-off.md) in your keymap and then invoking it.

Reviewer: ```suggestion
For keyboards which do not have such hardware, using soft off is as simple as placing the [soft off behavior](../../behaviors/soft-off.md) in your keymap and then invoking it.
```

You can then wake up the keyboard by pressing the reset button once, and repeating this for each side for split keyboards.

## Adding Soft Off to a Keyboard

More advanced methods of adding soft off to a keyboard are detailed below. The first two tabs describe methods involving hardware changes, while the last describes the firmware changes necessary to define a single specific keyswitch for waking up.

<Tabs groupId="advanced-methods" defaultValue="direct">
  <TabItem value="direct" label="Direct Pin"></TabItem>
  <TabItem value="matrix" label="Matrix Integrated Pin"></TabItem>
  <TabItem value="wakeup" label="Wakeup-only Keyswitch">
    The exact method of reusing a keyswitch to wake from the soft off state
    differs depending on whether said keyswitch is part of a direct GPIO kscan
    or part of a matrix kscan.
    <Tabs groupId="kscan-type" defaultValue="mkscan">
      <TabItem value="dkscan" label="Direct GPIO kscan"></TabItem>
      <TabItem value="mkscan" label="Matrix kscan"></TabItem>
    </Tabs>
  </TabItem>
</Tabs>

> ### Hardware Changes

Reviewer: ```suggestion

### Hardware Changes
```

    Ideally the switches used should be driven by the same matrix output pin so that both will be active simultaneously on the AND gate inputs. The alternative is to connect the switch to two MOSFETs that trigger both the regular matrix connect and the connect to the AND gate to ensure both pins are active/high at the same time even if scanning sets them high at different times.

  </TabItem>
  <TabItem value="wakeup" label="Wakeup-only Keyswitch">
    No hardware changes are necessary for this approach.
  </TabItem>
</Tabs>

### Firmware Changes

Several items work together to make both triggering soft off properly, and setting up the device to _wake_ from soft off work as expected.

<Tabs groupId="advanced-methods" defaultValue="direct" className="secrettabs">
  <TabItem value="direct" label="Direct Pin">
    <SoftOffBehaviour />
  </TabItem>
  <TabItem value="matrix" label="Matrix Integrated Pin">
    <SoftOffBehaviour />
  </TabItem>
  <TabItem value="wakeup" label="Wakeup-only Keyswitch">

> #### Soft off behavior

Reviewer: ```suggestion

#### Soft off behavior
```

      <TabItem value="dkscan" label="Direct GPIO kscan">
        <SidebandWakeupDirect />
      </TabItem>
      <TabItem value="mkscan" label="Matrix kscan">
        <SidebandWakeupMatrix />
      </TabItem>
    </Tabs>
  </TabItem>
</Tabs>

Here are the properties for the kscan sideband behaviors node:

<Tabs groupId="advanced-methods" defaultValue="direct" className="secrettabs">
  <TabItem value="direct" label="Direct Pin">
    <PropsSidebandDirect />
    <PropsSidebandChildren />
  </TabItem>
  <TabItem value="matrix" label="Matrix Integrated Pin">
    <PropsSidebandMatrix />
    <PropsSidebandChildren />
  </TabItem>
  <TabItem value="wakeup" label="Wakeup-only Keyswitch">
    <Tabs groupId="kscan-type" defaultValue="mkscan" className="secrettabs">
      <TabItem value="dkscan" label="Direct GPIO kscan">
        <PropsSidebandDirect />
      </TabItem>
      <TabItem value="mkscan" label="Matrix kscan">
        <PropsSidebandMatrix />
      </TabItem>
    </Tabs>
  </TabItem>
</Tabs>

<Tabs groupId="advanced-methods" defaultValue="direct" className="secrettabs">
  <TabItem value="direct" label="Direct Pin"></TabItem>
  <TabItem value="matrix" label="Matrix Integrated Pin">

>   You also need to update the `zmk,kscan` chosen value to point to the new kscan instance:

Reviewer: ```suggestion
    You also need to update the `zmk,kscan` chosen value to point to the new kscan instance:
```

```
/ {
    chosen {
        ...
        zmk,kscan = &side_band_behavior_triggers;
        ...
    };
};
```

>   <SoftOffWaker />

Reviewer: ```suggestion
    <SoftOffWaker />
```

  </TabItem>
  <TabItem value="wakeup" label="Wakeup-only Keyswitch">
    <Tabs groupId="kscan-type" defaultValue="mkscan" className="secrettabs">
        <TabItem value="dkscan" label="Direct GPIO kscan"></TabItem>
        <TabItem value="mkscan" label="Matrix kscan"><SoftOffWaker /></TabItem>
    </Tabs>
  </TabItem>
</Tabs>

Finally, we will list the `wakeup_scan` device in an additional configuration section so that the ZMK soft off process knows it needs to enable this device as part of the soft off processing so it can wake the keyboard from soft off when pressed:

```
/ {
    soft_off_wakers {
        compatible = "zmk,soft-off-wakeup-sources";
        wakeup-sources = <&wakeup_scan>;
    };
};
```

Here are the properties for the node:

- The `compatible` property for the node must be `zmk,soft-off-wakeup-sources`.
- The `wakeup-sources` property is a [phandle array](../../config/index.md#devicetree-property-types) pointing to all the devices that should be enabled during the shutdown process to be sure they can later wake the keyboard.

> :::tip

Reviewer: Should we put this under "wakeup-only" tab? It is essentially wakeup-only but with any key, isn't it?

Author: We certainly could and it would make sense. I quite like it where it is because it gives a bit better intuition of what `wakeup-sources` does, and I don't think it's causing any confusion at that location.
